lib/chat_notifier.rb

```diff
@@ -80,36 +80,67 @@ def enqueue_mentioned_job(user_ids)
     Jobs.enqueue_in(3.seconds, :create_chat_mention_notifications, {
       chat_message_id: @chat_message.id,
       user_ids: user_ids,
+      user_ids_to_group_mention_map: user_ids_to_group_mention_map,
       timestamp: @timestamp
     })
   end
 
-  def cooked_mentions
-    @cooked_mentions ||= Nokogiri::HTML5.fragment(@chat_message.cooked).css(".mention").map(&:text)
+  def direct_mentions_from_cooked
+    @direct_mentions_from_cooked ||= Nokogiri::HTML5.fragment(@chat_message.cooked).css(".mention").map(&:text)
+  end
+
+  def group_name_mentions
+    @group_mentions_from_cooked ||= Nokogiri::HTML5.fragment(@chat_message.cooked).css(".mention-group").map(&:text).map { |m| m[1..-1] }
+  end
+
+  def group_mentioned_users
+    @group_mentioned_users ||= begin
+                                 if group_name_mentions.empty?
+                                   return []
+                                 else
+                                   mentionable_groups = Group.mentionable(@user, include_public: false).where(name: group_name_mentions)
+                                   users_preloaded_query.joins(:groups).where(groups: mentionable_groups)
+                                 end
+                               end
+  end
+
+  def user_ids_to_group_mention_map
+    map = {}
+    group_mentioned_users.each do |user|
+      group_name = (user.groups.map(&:name) & group_name_mentions).first
+      map[user.id] = group_name
+    end
+    map
```

Comment:
The method looks funky, but it avoids doing 1 query for each group mentioned; groups are preloaded in the user query so we can loop through users from 1 query, and get the matching group name mentioned.